### ISSUE TITLE
As a website visitor I want to have a usable website even after my session expired or the UI was closed

### ISSUE THREAD

**Issue by paulroemer** [vaadin/flow#2058]
As a website visitor it's likely that your session will expire or the UI will be closed or the client-side is disconnected because of other reasons. As we disabled all kinds of confusing notifications the user will never know about the website's state.

Unfortunately, this also means that the user expects the page to be fully functional. So, he/she may try to login and presses buttons. But nothing will happen and the only information you will get is in the dev console: ```Trying to invoke method on not yet started or stopped application```. Which means that the application knows that something is not wrong but just accepts it.

Can't we just try to reconnect to the server or trigger a page reload. Same behavior as if you would press a (router) link.

**Comment by jouni** [comment]
Continuing the brainstorm above, if the developer could configure when to show the message, I think it would be great if the developer could also configure _where_ to show the message. For example, if a user was filling a feedback form, comes back after the session is expired and tries to submit the form, it would probably be nicer to show the message right next to the form instead of a banner at the top of the viewport.

**Comment by heruan** [comment]
In my experience with FW8, most of the times the "Session Expired" popup was shown a silent new session wouldn't had hurt and could have saved the user from a forced reload. Also, we have apps displayed on video-walls to show charts and dashboards, where there is literally no user (and no keyboard) so when a session expires we need to log on the video-wall system and force a reload on every screen, where a silent new session would be fine.

**Comment by pleku** [comment]
I think we will want to design and fix this properly before releasing 1.0. It can have a major impact for UX thus I consider it to be critical we make this as customizable as possible, with sensible (and good looking) default behavior.

Now just need to make the designs and find the time for the implementation.

**Comment by jouni** [comment]
@pleku, any update on this? Is this still a priority for 1.0?

I’m still up for helping with the design, but I can’t do it alone.

**Comment by pleku** [comment]
It is not currently a priority for 1.0, unfortunately.

We should do the design at some point though so then we could evaluate this as an enhancement to do during maintenance.

**Comment by pleku** [comment]
Decided to take this for Flow 1.0 scope and change the default to reload the web page on _any_ user interaction for the following use cases:
- Internal Error Dialog (client side exception)
- Session Expired Dialog (Session has expired)

Currently the session expired dialog is triggered by default and show a message to the user, and can be only triggered off by removing the messages.
Currently the session expired dialog is triggered on client side when the user interacts with the UI in a way that causes a server side request be done.

The defaults are changed for both dialogs in the PRs #3665 and #3666.
We will investigate if we should also change the session expired to happen immediately on _any_ user interaction on the page, instead of just something that causes a server communication.

**Comment by mstahv** [comment]
Please do those as defaults. This is also a developer experience issue. Then when developing with e.g. Spring dev tools, the view would be automatically visible on a second screen if you do a change in your IDE. //cc: @samie

**Comment by jouni** [comment]
I’ve spent some time thinking about various use cases around this, and trying to visualize them somehow to get a better understanding of the whole picture.

I ended with a flow diagram (no pun intended) for three use cases, which are the archetypal I identified the last one is not necessarily even directly related to Flow sessions, but it seemed relevant in this context. Feel free to add and disagree with these.

I tried not to be limited by technology and concentrated more on keeping the end user experience in tack. I think I ended up with a proposal that might be technically unfeasible, but I need @pleku, @Legioth and @Artur- to answer that.

I also didn’t yet think what kind of API changes this might mean, what would be left as the developer’s responsibility.

I think the current Flow behavior is somewhere between cases 1 and 2, and that’s the reason why it feels odd.

![](https://user-images.githubusercontent.com/66382/38408176-0e2eb920-3985-11e8-9d01-1c75180183a6.png)

Diagram source: https://drive.google.com/file/d/1nG4VMtffwD2Yzjx-rso-JmgPkIN49TJu/view?usp=sharing

A more rough ideation document I used (in case you want some more food for thought): https://www.figma.com/file/aipLqwnRqAGaWqWx5VAIaDBi/Flow-system-message-scenarios

PS. I noticed while testing the behavior on ticketmaster.com, that Safari seems to stop client-side timers/rendering in some cases when the page is in the background. Possibly something to keep in mind when designing the implementation.

**Comment by masbaehr** [comment]
For those who want a scenario like in "case 2". (E.g. on session expiration the page returns to the login screen) this should do as a workaround

```properties
server.servlet.session.timeout=5m
```
In a global js file put this. 
```javascript
var reloadTimer;
var sessionTimeoutTime = (5*1000*60)+500;
document.addEventListener("DOMContentLoaded", function(event) { 
   reloadTimer = setTimeout(function(){
    Vaadin.Flow.clients.ROOT.poll();
   }, sessionTimeoutTime);  
});
(function() { // Overriding XMLHttpRequest
    var oldXHR = window.XMLHttpRequest;
    function newXHR() {
        var realXHR = new oldXHR();
        realXHR.addEventListener("readystatechange", function() { 
            //console.log("an ajax request was made");
            clearTimeout(reloadTimer);
            reloadTimer = setTimeout(function(){
                Vaadin.Flow.clients.ROOT.poll();
            }, sessionTimeoutTime);
        }, false);

        return realXHR;
    }
    window.XMLHttpRequest = newXHR;
})();
```
Feel free to improve/correct my workaround, or show other ways to do it, as I'm not yet happy with this.. feels "hacky" ;)

**Comment by mrts** [comment]
Would saving current state (data entered etc) to [sessionStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage) before page reload and restoring it from there after reload perhaps be a way out?

**Comment by jouni** [comment]
@mrts yeah, for “case 1” that would one implementation alternative. I think it implies a full page reload, so it’s slightly less convenient for the end user. But it would be a great improvement.
